src/libstore/local-overlay-store.md

```diff
@@ -28,20 +30,21 @@ The parts of a local overlay store are as follows:
   The lower store must not change while it is mounted as part of an overlay store.
   To ensure it does not, you might want to mount the store directory read-only (which then requires the [read-only] parameter to be set to `true`).
 
-  Specified with the [`lower-store`](#store-experimental-local-overlay-store-lower-store) setting.
-
   - **Lower store directory**:
 
-    This is the directory used/exposed by the lower store.
+    > Specified with `lower-store.real` setting.
```

Comment:
I'm still a bit puzzled by what `.real` is and whether it's separate from `lower-store`, and whether it is an internal detail or one exposed to the user. A usage example might help elucidate this.

Reply:
This does improve things a bit for me.

This sentence confuses things for me, and suggests I'm missing something:
> "depending on the type of lower store chosen"

I presume this refers to: https://nixos.org/manual/nix/unstable/store/types/

But I can't see how any type of lower store makes sense other than the local store, other than perhaps the `mounted-ssh-store`, or some daemon socket. Again I wonder if some usage examples would help.

I'm interested in the usecase of providing a fuse or automounter filesystem as the lower layer which auto-unpacks nar files from a binary cache on access for example. Which it seems like I should be able to achieve, but I'm a little unclear on how you might provide the metadata for this concept, or if it's even possible/straightforward without significant additional changes to nix.

Reply:
> I'm still a bit puzzled by what `.real` is and whether it's separate from lower-store

The current syntax sucks: local store takes another URL, so what we have is a "query param within a query param". I hope there will be better syntax for this eventually. See https://github.com/NixOS/nix/issues/8373

> other than perhaps the `mounted-ssh-store`, or some daemon socket.

Yes, those are valid other options!

> I'm interested in the usecase of providing a fuse or automounter filesystem as the lower layer which auto-unpacks nar files from a binary cache on access for example. Which it seems like I should be able to achieve, but I'm a little unclear on how you might provide the metadata for this concept, or if it's even possible/straightforward without significant additional changes to nix.

It's very doable. Internally the lower story is any store which subclasses `LocalFSStore ` https://hydra.nixos.org/build/256242947/download/1/html/classnix_1_1LocalFSStore.html, so you can implement the missing methods however you want. One can also reuse `unix://` / `mounted-ssh-ng://` as clients (no new C++) and implement the daemon protocol.

Reply:
I still don't understand what `lower-store.real` is.  There are 0 hits for it in the codebase. can't find this option in https://github.com/NixOS/nix/blob/master/src/libstore/local-overlay-store.hh either.

Reply:
ah it's a query parameter of `lower-store`